client/components/Editor/utils/renderStatic.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * renderStatic() turns a list of ProseMirror nodes (in dehydrated JSON form) into a list of React
+ * elements. This is useful for creating HTML from documents on the server side, since ProseMirror's
+ * standard tools for doing this rely on the presence of a DOM API in the global scope.
+ *
+ * There are two steps to this process:
+ *
+ * 1. Convert Prosemirror nodes to the "output spec" format generated by Prosemirror toDOM methods.
+ *    This is an HTML-equivalent representation made of nested arrays.
+ * 	  (See here: https://prosemirror.net/docs/ref/version/0.18.0.html#model.DOMOutputSpec)
+ * 2. Turn the output spec into React elements.
+ */
+import React from 'react';
+import css from 'css';
+import camelCaseCss from 'camelcase-css';
+
+const parseStyleToObject = (style) => {
+	try {
+		const styleObj = {};
+		const wrappedStyle = `* { ${style} } `;
+		const cssAst = css.parse(wrappedStyle);
+		const { declarations } = cssAst.stylesheet.rules[0];
+		declarations.forEach(({ property, value }) => {
+			const camelCaseProperty = camelCaseCss(property);
+			styleObj[camelCaseProperty] = value;
+		});
+		return styleObj;
+	} catch (_) {
+		return {};
+	}
+};
+
+const attrsTransformations = {
+	rowspan: 'rowSpan',
+	colspan: 'colSpan',
+	class: 'className',
+	style: (val) => {
+		return { style: typeof val === 'string' ? parseStyleToObject(val) : val };
+	},
+};
+
+const normalizeAttrs = (attrs) => {
+	let resAttrs = {};
+	Object.entries(attrs).forEach(([key, value]) => {
+		if (key in attrsTransformations) {
+			const transformValue = attrsTransformations[value];
+			if (typeof transformValue === 'function') {
+				resAttrs = { ...resAttrs, ...transformValue(value) };
+			} else {
+				resAttrs[transformValue] = value;
+			}
+		} else {
+			resAttrs[key] = value;
+		}
+	});
+	return resAttrs;
+};
+
+const getAttrsFromOutputSpec = (maybeAttrs) => {
+	const hasAttrs = maybeAttrs && typeof maybeAttrs === 'object' && !Array.isArray(maybeAttrs);
+	const attrs = hasAttrs ? normalizeAttrs(maybeAttrs) : {};
+	return { attrs: attrs, hasAttrs: hasAttrs };
+};
+
+const createReactFromOutputSpec = (outputSpec, outermostKey) => {
+	const createReactInner = (spec, key) => {
+		if (!Array.isArray(spec)) {
+			return spec;
+		}
+		const [tagName, maybeAttrs, ...restItems] = spec;
+		const { attrs, hasAttrs } = getAttrsFromOutputSpec(maybeAttrs);
+		const children = hasAttrs ? restItems : [maybeAttrs, ...restItems];
+		return React.createElement(
+			tagName,
+			{ ...attrs, key: key },
+			...children.map((child, index) => createReactInner(child, `${key}-${index}`)),
+		);
+	};
```

Comment:
Just out of curiosity, why do we have `createReactInner ` inside of an outer function, when the inner function has no dependencies on the parent scope and recursion is handled internally?

Reply:
hahaha...absolutely no reason. this outer function used to close over a third variable which I since removed, but I forgot to simplify things here. Nice catch!